File: latest/workshop1_m2/W01_M02_demo.py
"""
============================================================
Workshop 1 — Module 2: Tufte's Principles
Python Demonstration Script
Data Visualization for Data Scientists — UNYT Tirana
============================================================
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os

os.makedirs("output", exist_ok=True)
plt.rcParams["figure.dpi"] = 150


# ── 1. Data-Ink Ratio: Progressive Erasure ──────────────────
cats = ["Compact", "Midsize", "SUV", "Pickup", "Minivan", "Subcompact", "2seater"]
vals = [47, 41, 62, 33, 11, 35, 5]

fig, axes = plt.subplots(1, 4, figsize=(16, 4))

# V1: Default matplotlib (heavy styling)
ax = axes[0]
ax.bar(cats, vals, color="steelblue", edgecolor="black", linewidth=1)
ax.set_facecolor("#EEEEEE")
ax.grid(True, linewidth=0.8, color="white")
ax.set_title("V1: Default", fontsize=9, fontweight="bold")
ax.tick_params(labelsize=6)
ax.set_xticklabels(cats, rotation=50, ha="right")

# V2: Remove background and border
ax = axes[1]
ax.bar(cats, vals, color="steelblue", edgecolor="none")
ax.grid(axis="y", linewidth=0.3, color="#DDDDDD")
ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
ax.set_title("V2: No Fill/Border", fontsize=9, fontweight="bold")
ax.tick_params(labelsize=6)
ax.set_xticklabels(cats, rotation=50, ha="right")

# V3: Horizontal, no grid
ax = axes[2]
sorted_idx = np.argsort(vals)
s_cats = [cats[i] for i in sorted_idx]
s_vals = [vals[i] for i in sorted_idx]
ax.barh(s_cats, s_vals, color="steelblue", height=0.6, edgecolor="none")
ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
ax.set_title("V3: Horizontal, No Grid", fontsize=9, fontweight="bold")
ax.tick_params(labelsize=7)

# V4: Tufte style
ax = axes[3]
ax.barh(s_cats, s_vals, color="#1565C0", height=0.5, edgecolor="none")
for i, v in enumerate(s_vals):
    ax.text(v + 0.8, i, str(v), va="center", fontsize=8, color="#333")
for sp in ["top", "right", "left", "bottom"]:
    ax.spines[sp].set_visible(False)
ax.set_xticks([])
ax.tick_params(left=False, labelsize=7)
ax.set_title("V4: Tufte Style", fontsize=9, fontweight="bold")

fig.suptitle("Data-Ink Ratio: Progressive Erasure (Python)",
             fontsize=12, fontweight="bold")
plt.tight_layout()
plt.savefig("output/data_ink_erasure_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 2. Chartjunk: Pie vs Clean Bar ─────────────────────────
labels = ["Product A", "Product B", "Product C", "Product D"]
sizes = [35, 30, 20, 15]

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 4))

# Pie chart (demonstration of what to avoid)
ax1.pie(sizes, labels=labels, autopct="%1.0f%%", shadow=True,
        colors=["#FF6B6B", "#4ECDC4", "#FFE66D", "#DDA0DD"],
        explode=(0.05, 0, 0, 0), textprops={"fontsize": 8})
ax1.set_title("Pie Chart\n(hard to compare slices)", fontsize=9,
              fontweight="bold", color="#C62828")

# Clean bar
idx = np.argsort(sizes)
s_labels = [labels[i] for i in idx]
s_sizes = [sizes[i] for i in idx]
ax2.barh(s_labels, s_sizes, color="#1565C0", height=0.55)
for i, v in enumerate(s_sizes):
    ax2.text(v + 0.5, i, f"{v}%", va="center", fontsize=9)
ax2.set_xlabel("Market Share (%)", fontsize=9)
ax2.set_title("Horizontal Bar\n(easy to compare lengths)", fontsize=9,
              fontweight="bold", color="#1565C0")
ax2.spines["top"].set_visible(False)
ax2.spines["right"].set_visible(False)
ax2.grid(axis="x", alpha=0.15)
ax2.tick_params(labelsize=8)

fig.suptitle("Chartjunk Removal: Pie → Bar", fontsize=12, fontweight="bold")
plt.tight_layout()
plt.savefig("output/chartjunk_removal_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 3. Lie Factor: Truncated vs Full Axis ───────────────────
quarters = ["Q1", "Q2", "Q3", "Q4"]
revenue = [982, 995, 1003, 1010]

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 4))

# Truncated axis (misleading)
ax1.bar(quarters, revenue, color="#E53935", width=0.5, edgecolor="white")
ax1.set_ylim(975, 1015)
ax1.set_title("Truncated Axis\n\"Revenue soaring!\"",
              fontsize=9, fontweight="bold", color="#C62828")
ax1.set_ylabel("Revenue ($M)", fontsize=8)
ax1.tick_params(labelsize=7)

# Full axis (honest)
ax2.bar(quarters, revenue, color="#1565C0", width=0.5, edgecolor="white")
ax2.set_ylim(0, 1200)
for i, v in enumerate(revenue):
    ax2.text(i, v + 20, f"${v}M", ha="center", fontsize=7)
ax2.set_title("Full Axis\n\"Revenue stable\"",
              fontsize=9, fontweight="bold", color="#1565C0")
ax2.set_ylabel("Revenue ($M)", fontsize=8)
ax2.spines["top"].set_visible(False)
ax2.spines["right"].set_visible(False)
ax2.tick_params(labelsize=7)

fig.suptitle("Lie Factor: Baseline Matters", fontsize=12, fontweight="bold")
plt.tight_layout()
plt.savefig("output/lie_factor_demo_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 4. Small Multiples ─────────────────────────────────────
np.random.seed(7)
regions = ["North", "South", "East", "West", "Central", "Coastal"]
months = np.arange(1, 13)

fig, axes = plt.subplots(2, 3, figsize=(10, 5), sharey=True)

for ax, region in zip(axes.flatten(), regions):
    base = np.random.uniform(50, 120)
    trend = np.cumsum(np.random.normal(2, 8, 12)) + base

    ax.plot(months, trend, color="#1565C0", linewidth=1.5)
    ax.fill_between(months, trend, alpha=0.08, color="#1565C0")
    ax.set_title(region, fontsize=9, fontweight="bold")
    ax.set_xticks([1, 6, 12])
    ax.set_xticklabels(["Jan", "Jun", "Dec"], fontsize=6)
    ax.tick_params(labelsize=6)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.grid(axis="y", alpha=0.15)

fig.suptitle("Small Multiples: Same Scale, Easy Comparison (Python)",
             fontsize=11, fontweight="bold")
plt.tight_layout()
plt.savefig("output/small_multiples_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 5. Sparklines ──────────────────────────────────────────
np.random.seed(12)
metrics = ["Revenue", "Users", "Retention", "NPS Score"]

fig, axes = plt.subplots(4, 1, figsize=(5, 4))

for ax, metric in zip(axes, metrics):
    data = np.cumsum(np.random.normal(0.5, 3, 52))
    ax.plot(data, color="#1565C0", linewidth=1.0)
    ax.fill_between(range(len(data)), data, alpha=0.04, color="#1565C0")

    # Mark min (red), max (green), endpoint (blue)
    ax.plot(np.argmin(data), data.min(), "o", color="#E53935", markersize=4)
    ax.plot(np.argmax(data), data.max(), "o", color="#2E7D32", markersize=4)
    ax.plot(len(data) - 1, data[-1], "o", color="#1565C0", markersize=3)

    # Remove ALL non-data elements
    ax.set_xticks([])
    ax.set_yticks([])
    for sp in ax.spines.values():
        sp.set_visible(False)

    ax.set_ylabel(metric, fontsize=8, rotation=0, labelpad=55, va="center")

fig.suptitle("Sparklines: Word-Sized Graphics (Tufte, 2006)",
             fontsize=10, fontweight="bold")
plt.tight_layout()
plt.savefig("output/sparklines_py.png", dpi=150, bbox_inches="tight")
plt.close()


print("── All M02 plots saved to output/ ──")

File: latest/week1/workshop1_m2/W01_M02_demo.R
# ============================================================
# Workshop 1 — Module 2: Tufte's Principles
# R Demonstration Script
# Data Visualization for Data Scientists — UNYT Tirana
# ============================================================

# ── 0. Setup ────────────────────────────────────────────────
# install.packages(c("tidyverse", "ggthemes", "patchwork"))
library(tidyverse)
library(ggthemes)   # theme_tufte(), theme_economist(), etc.
library(patchwork)

# Create output directory
dir.create("output", showWarnings = FALSE)


# ── 1. Data-Ink Ratio: Progressive Erasure ──────────────────
# Using the mpg dataset to demonstrate four stages of
# chartjunk removal following Tufte's "erase" principle.

mpg_counts <- mpg |>
  count(class) |>
  mutate(class = fct_reorder(class, n))

# Version 1: Default ggplot2 (heavy gridlines, grey background)
p1 <- ggplot(mpg_counts, aes(x = class, y = n)) +
  geom_bar(stat = "identity", fill = "skyblue", color = "black") +
  labs(title = "V1: Default", y = "Count", x = NULL) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 7),
        plot.title = element_text(size = 9, face = "bold"))

# Version 2: Remove background fill and border
p2 <- ggplot(mpg_counts, aes(x = class, y = n)) +
  geom_col(fill = "steelblue") +
  labs(title = "V2: No Fill/Border", y = "Count", x = NULL) +
  theme_minimal(base_size = 9) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 7),
        plot.title = element_text(face = "bold"))

# Version 3: Remove gridlines, flip to horizontal
p3 <- ggplot(mpg_counts, aes(x = class, y = n)) +
  geom_col(fill = "steelblue", width = 0.6) +
  coord_flip() +
  labs(title = "V3: Horizontal, No Grid", y = "Count", x = NULL) +
  theme_minimal(base_size = 9) +
  theme(panel.grid.major.y = element_blank(),
        panel.grid.minor = element_blank(),
        plot.title = element_text(face = "bold"))

# Version 4: Tufte-style — direct labels, minimal spines
p4 <- ggplot(mpg_counts, aes(x = class, y = n)) +
  geom_col(fill = "#1565C0", width = 0.5) +
  geom_text(aes(label = n), hjust = -0.3, size = 3) +
  coord_flip() +
  theme_tufte(base_size = 9) +
  labs(title = "V4: Tufte Style", y = NULL, x = NULL) +
  theme(plot.title = element_text(face = "bold"))

# Combine with patchwork
p_combined <- (p1 | p2) / (p3 | p4) +
  plot_annotation(
    title = "Data-Ink Ratio: Progressive Erasure",
    subtitle = "Same data, four design stages"
  )

print(p_combined)
ggsave("output/data_ink_erasure.png", p_combined,
       width = 10, height = 7, dpi = 150)


# ── 2. Chartjunk: 3-D Pie vs Clean Bar ─────────────────────
# R does not natively support 3-D pies, which is itself a
# design endorsement. Compare a standard pie to a bar chart.

share_data <- tibble(
  product = c("Product A", "Product B", "Product C", "Product D"),
  share = c(35, 30, 20, 15)
) |>
  mutate(product = fct_reorder(product, share))

# Pie (not recommended, but shown for comparison)
p_pie <- ggplot(share_data, aes(x = "", y = share, fill = product)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar("y") +
  theme_void() +
  labs(title = "Pie Chart (Hard to Compare)") +
  theme(plot.title = element_text(size = 9, face = "bold"))

# Clean horizontal bar
p_bar <- ggplot(share_data, aes(x = product, y = share)) +
  geom_col(fill = "#1565C0", width = 0.6) +
  geom_text(aes(label = paste0(share, "%")), hjust = -0.2, size = 3.5) +
  coord_flip() +
  theme_tufte(base_size = 10) +
  labs(title = "Tufte Bar (Easy to Compare)",
       x = NULL, y = "Market Share (%)") +
  theme(plot.title = element_text(face = "bold"))

p_chartjunk <- p_pie + p_bar +
  plot_annotation(title = "Chartjunk Removal: Pie → Bar")

print(p_chartjunk)
ggsave("output/chartjunk_removal.png", p_chartjunk,
       width = 10, height = 4, dpi = 150)


# ── 3. Lie Factor Demonstration ─────────────────────────────
# Show truncated vs full y-axis for the same data

revenue <- tibble(
  quarter = c("Q1", "Q2", "Q3", "Q4"),
  value = c(982, 995, 1003, 1010)
)

# Truncated (lie factor >> 1)
p_trunc <- ggplot(revenue, aes(x = quarter, y = value)) +
  geom_col(fill = "#E53935", width = 0.5) +
  coord_cartesian(ylim = c(975, 1015)) +
  labs(title = "Truncated Axis (Misleading)",
       y = "Revenue ($M)", x = NULL) +
  theme_minimal(base_size = 9) +
  theme(plot.title = element_text(face = "bold", color = "#C62828"))

# Honest (lie factor ~ 1)
p_honest <- ggplot(revenue, aes(x = quarter, y = value)) +
  geom_col(fill = "#1565C0", width = 0.5) +
  geom_text(aes(label = value), vjust = -0.5, size = 3) +
  coord_cartesian(ylim = c(0, 1200)) +
  labs(title = "Full Axis (Honest)",
       y = "Revenue ($M)", x = NULL) +
  theme_minimal(base_size = 9) +
  theme(plot.title = element_text(face = "bold", color = "#1565C0"))

p_lie <- p_trunc + p_honest +
  plot_annotation(title = "Lie Factor: Truncated vs Full Baseline")

print(p_lie)
ggsave("output/lie_factor_demo.png", p_lie,
       width = 9, height = 4, dpi = 150)


# ── 4. Small Multiples ─────────────────────────────────────
# Simulate regional monthly revenue data

set.seed(7)
regions <- c("North", "South", "East", "West", "Central", "Coastal")
months <- 1:12

sm_data <- expand_grid(region = regions, month = months) |>
  group_by(region) |>
  mutate(
    base = runif(1, 50, 120),
    revenue = cumsum(rnorm(n(), 2, 8)) + base
  ) |>
  ungroup()

p_sm <- ggplot(sm_data, aes(x = month, y = revenue)) +
  geom_line(color = "#1565C0", linewidth = 0.8) +
  geom_area(fill = "#1565C0", alpha = 0.08) +
  facet_wrap(~region, ncol = 3, scales = "fixed") +
  scale_x_continuous(breaks = c(1, 6, 12),
                     labels = c("Jan", "Jun", "Dec")) +
  theme_minimal(base_size = 9) +
  theme(
    strip.text = element_text(face = "bold"),
    panel.grid.minor = element_blank()
  ) +
  labs(x = NULL, y = "Revenue ($K)",
       title = "Small Multiples: Same Scale, Easy Comparison",
       subtitle = "\"Compared to what?\" — Tufte")

print(p_sm)
ggsave("output/small_multiples.png", p_sm,
       width = 9, height = 5, dpi = 150)


# ── 5. Sparklines (using theme_void) ───────────────────────
set.seed(12)
spark_data <- tibble(
  week = rep(1:52, 4),
  metric = rep(c("Revenue", "Users", "Retention", "NPS"),
               each = 52),
  value = c(
    cumsum(rnorm(52, 0.5, 3)),
    cumsum(rnorm(52, 0.3, 2)),
    cumsum(rnorm(52, 0.1, 4)),
    cumsum(rnorm(52, 0.4, 2.5))
  )
)

p_spark <- ggplot(spark_data, aes(x = week, y = value)) +
  geom_line(color = "#1565C0", linewidth = 0.6) +
  facet_wrap(~metric, ncol = 1, scales = "free_y") +
  theme_void(base_size = 8) +
  theme(strip.text = element_text(hjust = 0, face = "bold", size = 8)) +
  labs(title = "Sparklines: Word-Sized Graphics (Tufte, 2006)")

print(p_spark)
ggsave("output/sparklines.png", p_spark,
       width = 5, height = 4, dpi = 150)

cat("\n── All M02 plots saved to output/ ──\n")
